### PR TITLE
RestAPI

### PR DESCRIPTION
**Pull request by fullsushidev**
## Description
The basic structure of the RestAPI plugin with 3 endpoints.

## Motivation and context
This basic structure is intended to be used in our first demo session and the following endpoints were implemented:

> **Given a product and a timestamp, reconstruct its dependency network**
> Return: A set of revisions, along with an adjacency matrix
> 
> REST examples:
> `GET /api/mvn/org.slf4j:slf4j-api/deps/1233323123`
> `GET /api/pypi/numpy/deps/1233323123?transitive=true`

> **Given a product and a timestamp, retrieve its call graph**
> Use case: A user wants to run a custom analysis locally.
> Return: A JSON-serialized RevisionCallGraph
> 
> REST examples:
> `GET /api/mvn/org.slf4j:slf4j-api/cg/1233323123`
> `GET /api/pypi/numpy/deps/1233323123?transitive=true`

> **Given a product and a version, retrieve all known metadata**
> Return: All known metadata for a revision
> 
> REST examples:
> `GET /api/mvn/org.slf4j:slf4j-api/1.7.29`
> `GET /api/pypi/numpy/1.15.2`

[_list of all endpoints_](https://www.fasten-project.eu/view/wiki/WP2/Queries%20to%20be%20supported%20by%20the%20FASTEN%20server/)

From the Vert.X framework, we are using the [Web API contract](https://vertx.io/docs/vertx-web-api-contract/java/) to load an [OpenAPI specification](https://www.openapis.org/) file and deploy its endpoints.

## Testing
(locally as a stand-alone plugin)

- **Requirement:** Access to the Metadata DB instance or a Metadata DB dump.
- Uncomment those [3 dependencies from the rest api pom file](https://github.com/fasten-project/fasten/blob/b87a7f505294431f186221b3020ce595cc23a3d0/analyzer/restapi-plugin/pom.xml#L46#L61)
- Add the DB password to a env var: `export FASTEN_DBPASS=your_pass`
- Go to the folder `cd analyzer/restapi_plugin`
- Run the following command to install and execute the plugin:
    ```
    mvn clean install && mvn exec:java -Dexec.args='-d YOUR_DB_ADDRESS -u YOUR_USER'
    ```
- Once the server is up, curl the endpoint or just follow this link:
    [http://127.0.0.1:8080/api/mvn/au.org.consumerdatastandards:reflection/1.1.1](http://127.0.0.1:8080/api/mvn/au.org.consumerdatastandards.reflection/1.1.1)

_Note: the default value for the DB address and the user are the same from MetadataPlugin, respectively `jdbc:postgresql:postgres` and `postgres`_

## Task list  
- [x] Create a server to load the OpenAPI specification document and create the router (endpoints/errors handlers, etc)
- [x] Create a plugin to connect to the metadata DB and deploy the server
- [x] Create a CLI to connect to the DB and run the plugin
- [x] Check if the API server deployment needs to be updated to work within FASTEN server
- [x] Create a query to retrieve all known metadata for a product+verion
- [x] Update the endpoint to use the query above
- [x]  Update docs for each file

## Additional context
Currently the RestAPI outputs a JSON returned fetched from the JOOQ library, this data should be parsed to a human-friendly format as soon as the test data is available in the metadata database.

### CODE REVIEW

**Comment by senaingrid** [comment]
@mir-am Could you please take a look and give your first impressions? Thanks

**Comment by mir-am** [comment]
> @mir-am Could you please take a look and give your first impressions? Thanks

Thanks for adding a description and task list. For now, I've skimmed through the plug-in class and `Main`. It seems okay. It would be great if you could implement a query to see some data from the metadata DB. I think that you can use JOOQ to query the DB. Also, use the DB dump that you have for development.

**Comment by MihhailSokolov** [comment]
@fullsushidev Concerning moving `MetadataDao` class, I have already moved it to the core in my PR #34. So you can either just merge master into your branch and resolve any problems once my PR is done and merged (which I hope will happen in about a week), or you can see where I moved it and do the same to avoid any merge conflicts in the future :)

File: analyzer/restapi-plugin/README.md
Comment (ilyagrishkov) [inline analyzer/restapi-plugin/README.md:14]: I have added a template README. The descriptions of the plugin needs to be added before merging this PR.

File: analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/RestAPIPlugin.java
Comment (marcomicera) [reply analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/RestAPIPlugin.java]: > Yes, I think this is the right place to deploy the server. However, I would make it a private variable of the class, so that this server could be stopped from `stop()` method.

Done in 85c52c79.

**Comment by marcomicera** [comment]
> * Once the server is up, curl the endpoint or just follow this link:
>   http://127.0.0.1:8080/api/mvn/au.org.consumerdatastandards.reflection/1.1.1

Do you mean just http://127.0.0.1:8080/api? I get a 500 otherwise.

**Comment by marcomicera** [comment]
Rebasing to [`master`](https://github.com/fasten-project/fasten/tree/master)...

**Review by marcomicera — CHANGES_REQUESTED** [review]
I'm following the query examples [here](https://github.com/fasten-project/fasten/blob/0d48fdb29c6766605e2abb9ba0a4458d57463514/analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/server/OpenAPIServer.java) but I always get a 404: I should have a look at the DB content first...

File: analyzer/pom.xml
Comment (marcomicera) [reply analyzer/pom.xml:24]: The CI also fails for this very reason.

Comment (ilyagrishkov) [reply analyzer/pom.xml:24]: Indeed we removed this module quite some time ago. You can safely remove it here

**Comment by fullsushidev** [comment]
> > * Once the server is up, curl the endpoint or just follow this link:
> >   http://127.0.0.1:8080/api/mvn/au.org.consumerdatastandards.reflection/1.1.1
> 
> Do you mean just http://127.0.0.1:8080/api? I get a 500 otherwise.

No, I didn't. Please check if you're connected to the right DB dump and that it has this data.

File: analyzer/restapi-plugin/pom.xml
Comment (fullsushidev) [reply analyzer/restapi-plugin/pom.xml:62]: To avoid dependency duplication.
We already have those dependencies in the top-level pom files. So, to be able to run the plugin independently we need to declare those same dependencies here.
Also, we had some updates in the way the entire project is dealing with pom files, I'll double-check if this needs to be changed before making this PR ready to review.

File: analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/Main.java
Comment (fullsushidev) [reply analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/Main.java]: Actually, I just updated to:
`new CommandLine(new Main()).execute(args);`
Just like the POManalyzer, since the intention of this plugin is to persist.

**Comment by mir-am** [comment]
I've tested the REST API on the VM (FASTEN demo). The metadata of packages can be retrieved. However, I couldn't test the other endpoints since the input packages for the VM do not have timestamps.

**Comment by marcomicera** [comment]
> However, I couldn't test the other endpoints since the input packages for the VM do not have timestamps.

So our plan is to complete the logic behind those endpoints as soon as there are some timestamps in the DB. Could you please warn us as soon as this is done? A comment here should be enough. Thank you.

File: analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/server/OpenAPIServer.java
Comment (MihhailSokolov) [inline analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/server/OpenAPIServer.java:168]: I thought we agreed to retrieve a metadata per database entity (i.e. `/api/{forge}/{pkg_name}/{version}/callables/metadata`, `/api/{forge}/{pkg_name}/{version}/edges/metadata` and etc)?

**Comment by gousiosg** [comment]
This PR is good for testing the technologies and trying to deliver a quick prototype. However, we need to create a proper REST API (here we just have a couple of endpoints). I would like to suggest that we merge this ASAP and then start from the beginning with a proper RESTful design that exposes all DB entities. We can discuss how to do this elsewhere.

File: analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/server/OpenAPIServer.java
Comment (MihhailSokolov) [inline analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/server/OpenAPIServer.java:129]: It would be nice to have these handlers in separate methods. Having a method of 350 lines long is not very nice

File: analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/server/OpenAPIServer.java
Comment (MihhailSokolov) [inline analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/server/OpenAPIServer.java:400]: I don't think you actually need this method. It is used in Metadata plugin because some timestamps were in seconds instead of milliseconds and we want to always have milliseconds the database. Since you're only retrieving data from the database, you don't need to adjust any timestamps provided that input timestamp in milliseconds.

File: core/src/main/java/eu/fasten/core/data/metadatadb/MetadataDao.java
Comment (MihhailSokolov) [inline core/src/main/java/eu/fasten/core/data/metadatadb/MetadataDao.java:822]: I don't think that code with "dummy" queries should be merged. I understand that data is not available yet, just don't forget to implement it as soon as the data is there :)

**Review by MihhailSokolov — CHANGES_REQUESTED** [review]
Requested some changes. The Rest API plugin still needs some work especially what is missing is a way to "browse" the database. We should be able to retrieve each database entity with Rest API, for example, to retrieve all callables of a certain package like `junit:junit:4.1` version you should send request to `/api/mvn/junit/junit/4.12/callables` and so on.
Everything else, except some comments, looks good to me.

File: analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/RestAPIPlugin.java
Comment (fullsushidev) [reply analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/RestAPIPlugin.java:52]: I was following OPALplugin. Which plugin do you suggest for us to follow? Also considering all other changes.

Comment (MihhailSokolov) [reply analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/RestAPIPlugin.java:52]: Well, all other plugins just have strings like `"Metadata Plugin"`, `"Graph Plugin"`.
In the case of OPAL, it is fine because the class is also called `OPAL`, but here I think it would be nicer to have a `Rest API Plugin` (with spaces which will be absent in the case of this.getClass().getCanonicalName()``)

File: analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/server/OpenAPIServer.java
Comment (fullsushidev) [reply analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/server/OpenAPIServer.java:168]: Yes, in this case, the query was simplified to return the metadata for the package version, the other endpoints should be implemented as agreed in the last meeting.

Moreover, I agree with @gousiosg, this PR was intended to be a quick prototype showcasing canned queries as agreed in the Brussels meeting since the DB schema was still a work in progress.

To be able to move forward and implement a proper RESTful design we need clarification and the RESTful API Design Specification (deliverable 2.4)

**Comment by gousiosg** [comment]
Ok let's merge this one and move to a new, more comprehensive PR. 

@fullsushidev Please complete the review and then @ilyagrishkov or @MihhailSokolov please merge it.

**Review by marcomicera — APPROVED** [review]

**Review by ilyagrishkov — APPROVED** [review]

**Pull request merged**
